### ISSUE TITLE
Sentinel erronously reporting target for outbound push has expired TLS cert

### ISSUE THREAD

**Issue by mrjones-plip** [medic/cht-core#7957]
**Describe the bug**
I have set up an outbound push for a web server with a known good TLS cert.  However, the pushes fail with `Error: certificate has expired`

**To Reproduce**
1. Set up a CHT instance using [Docker Helper for 3.x](https://docs.communityhealthtoolkit.org/apps/guides/hosting/3.x/app-developer/#cht-docker-helper), be sure you're on latest release, `3.17.1`
2. add an outbound push, say for all patients
3. edit a patient to initiate a push
4. Instead of push working, there's an error in the sentinel logs `Error: certificate has expired`

**Expected behavior**
Outbound push should correctly work based on the `mapping` in the app settings.

**Logs**
```
tail -f /srv/storage/medic-sentinel/logs/medic-sentinel.log

[2022-12-02 21:15:15] 2022-12-02 21:15:15 ERROR: Failed to push ce84932c-858f-4f64-ba86-446346a82b1c to push-to-EXPIRED-tls-expired.badssl.com: Error: certificate has expired 
[2022-12-02 21:15:15] 2022-12-02 21:15:15 ERROR: Failed to push ce84932c-858f-4f64-ba86-446346a82b1c to push-to-VALID-tls-medic.org: Error: certificate has expired 
[2022-12-02 21:15:15] 2022-12-02 21:15:15 INFO: Task outbound completed 
```

**Screenshots**
NA

**Environment**
- Instance:  local dev instance
- Browser: Firefox latest
- Client platform: Ubuntu 20.04
- App: sentinel
- Version: CHT 3.17.1

**Additional context**
* Originally discovered in [this forum post](https://forum.communityhealthtoolkit.org/t/post-request-with-outbound/2197?u=mrjones).
* This does not affect 4.0.1, confirmed it failed correctly with `Failed to push fa5136f3-4adb-42c2-b945-6f1205edbe23 to push-to-EXPIRED-tls-expired.badssl.com: Error: certificate has expired` on  `expired.badssl.com` and then succeeded `INFO: Pushed fa5136f3-4adb-42c2-b945-6f1205edbe23 to push-to-VALID-tls-medic.org` to `medic.org`.
* not all valid, non-expired TLS certs seem to be affected, here's some test results:
   * VALID -  https://rapidpro.app.medicmobile.org (Gandi)
   * NOT VALID -  https://medic.org (Let's Encrypt) 
* Here's my app settings config - see below "outbound config from my `app_settings.json`"
*  Assuming this a node.js error, I [found that possibly Node 8 is affected](https://stackoverflow.com/a/62145446), but later versions aren't?  When I have some more time I'll research recreating this config in CHT 4.0.1 to see if the problem persists there. 
*  I also tried to update the `medic-os` packages with `apt upgrade && apt dist-upgrade`.  After which I restarted sentinel with `/boot/svc-restart medic-sentinel` and then edited another patient. This did not fix anything. Packages updated listed below.  - see "`apt dist-update` Packages" below.

----------

####  `apt dist-update` Packages: 
```
  apt apt-utils base-files ca-certificates curl distro-info-data dpkg git git-man krb5-locales libapt-inst2.0 libapt-pkg5.0 libc-bin libc6 libcurl3-gnutls libdns-export162 libglib2.0-0
  libglib2.0-data libgnutls30 libgssapi-krb5-2 libhogweed4 libisc-export160 libk5crypto3 libkrb5-3 libkrb5support0 libldap-2.4-2 libnettle6 libp11-kit0 libpam-modules libpam-modules-bin
  libpam-runtime libpam0g libperl5.22 libpython3.5 libpython3.5-minimal libpython3.5-stdlib libseccomp2 libssl1.0.0 libsystemd0 libudev1 libx11-6 libx11-data multiarch-support openssl perl
  perl-base perl-modules-5.22 python-apt-common python3-apt python3.5 python3.5-minimal screen sudo systemd systemd-sysv tar unattended-upgrades vim vim-common vim-runtime
```

#### outbound config from my `app_settings.json`
```yml
  "outbound": {
    "push-to-EXPIRED-tls-expired.badssl.com": {
      "relevant_to": "doc.type === 'person' && doc.role == 'patient'",
      "destination": {
        "base_url": "https://expired.badssl.com",
        "path": "/"
      },
      "mapping": {
        "resourceType": {
          "expr": "'Patient'"
        },
        "gender": "doc.sex",
        "name": {
          "expr": "[ { \"use\":\"official\", \"family\": doc.name , \"given\": [ doc.short_name ] } ]",
          "optional": false
        },
        "identifier.use": {
          "expr": "'usual'"
        },
        "identifier.value": {
          "expr": "'cht/'.concat(doc._id)"
        },
        "birthDate": "doc.date_of_birth",
        "telecom": {
          "expr": "[{ \"system\":\"phone\", \"value\": doc.phone, \"use\":\"mobile\" } ]",
          "optional": true
        },
        "active": {
          "expr": true,
          "optional": false
        }
      }
    },
    "push-to-VALID-tls-medic.org": {
      "relevant_to": "doc.type === 'person' && doc.role == 'patient'",
      "destination": {
        "base_url": "https://medic.org",
        "path": "/"
      },
      "mapping": {
        "resourceType": {
          "expr": "'Patient'"
        },
        "gender": "doc.sex",
        "name": {
          "expr": "[ { \"use\":\"official\", \"family\": doc.name , \"given\": [ doc.short_name ] } ]",
          "optional": false
        },
        "identifier.use": {
          "expr": "'usual'"
        },
        "identifier.value": {
          "expr": "'cht/'.concat(doc._id)"
        },
        "birthDate": "doc.date_of_birth",
        "telecom": {
          "expr": "[{ \"system\":\"phone\", \"value\": doc.phone, \"use\":\"mobile\" } ]",
          "optional": true
        },
        "active": {
          "expr": true,
          "optional": false
        }
      }
    }
  }
```

**Comment by mrjones-plip** [comment]
I was trying to get an MVP to reproduce the bug using the node version in the `medic-os` container, but have only failed.  so far I have:

* Find the source of `POST` call in sentinel, [I think it's here](https://github.com/medic/cht-core/blob/master/shared-libs/outbound/src/outbound.js#L183)! now let's isolate it...
* get shell: `docker exec -it medic-os bash`
* Find node, it's here! `/srv/software/medic-core/v2.1.1/x64/bin/node`
* symlink node: `ln -s /srv/software/medic-core/v2.1.1/x64/bin/node /usr/bin/node`
* Install some stuff: `/srv/software/medic-core/v2.1.1/x64/bin/npm install request-promise-native url-join`
* try and run a script, but get errors b/c I'm still not a JS dev :disappointed:  

```js
const urlJoin = require('url-join');
const request = require('request-promise-native');
const authOptions = {
  form: {
    login: 'foo',
    password: 'bar'
  },
  url: urlJoin('https://medic.org/', '/'),
  json: true
};

request.post(authOptions)
  .then(result => {
    // No that's not a spelling mistake, this API is sometimes French!
    if (result.statut !== 200) {
      console.log('Non-200 status from Muso auth: %o', result);
    }
  });
```

**Comment by mrjones-plip** [comment]
Aha - a work around!  I did the following to install node :

1. get shell: `docker exec -it medic-os bash`
1. ensure node isn't isntalled: `apt update && sudo apt-get remove nodejs`
2. add the apt repo with node 14: `curl -fsSL https://deb.nodesource.com/setup_14.x | sudo -E bash -`
3. install node (and friends): `sudo apt-get install nodejs gcc g++ make`
4. remove medic-os's node: `rm /srv/software/medic-core/v2.1.1/x64/bin/node`
5. symlink in the new node: `ln -s /usr/bin/node /srv/software/medic-core/v2.1.1/x64/bin/node`
6. reboot sentinel: `/boot/svc-restart medic-sentinel`

this both resulted in Node 14:

```
root@a1dec31e7a7f:/srv# /srv/software/medic-core/v2.1.1/x64/bin/node -v
v14.21.1
```

and the sentinel transitions now works:

```
INFO: Pushed ce84932c-858f-4f64-ba86-446346a82b1c to push-to-VALID-tls-medic.org
```

**Comment by magp18** [comment]
> Aha - a work around! I did the following to install node :
> 
> 1. get shell: `docker exec -it medic-os bash`
> 2. ensure node isn't isntalled: `apt update && sudo apt-get remove nodejs`
> 3. add the apt repo with node 14: `curl -fsSL https://deb.nodesource.com/setup_14.x | sudo -E bash -`
> 4. install node (and friends): `sudo apt-get install nodejs gcc g++ make`
> 5. remove medic-os's node: `rm /srv/software/medic-core/v2.1.1/x64/bin/node`
> 6. symlink in the new node: `ln -s /usr/bin/node /srv/software/medic-core/v2.1.1/x64/bin/node`
> 7. reboot sentinel: `/boot/svc-restart medic-sentinel`
> 
> this both resulted in Node 14:
> 
> ```
> root@a1dec31e7a7f:/srv# /srv/software/medic-core/v2.1.1/x64/bin/node -v
> v14.21.1
> ```
> 
> and the sentinel transitions now works:
> 
> ```
> INFO: Pushed ce84932c-858f-4f64-ba86-446346a82b1c to push-to-VALID-tls-medic.org
> ```

This works well! Thank you so much!

**Comment by mrjones-plip** [comment]
@magp18  - careful with the workaround though - it is ephemeral and changes will be lost when you restart the container.  A better solution, if possible, is to upgrade to CHT 4.

**Comment by garethbowen** [comment]
Update from call this morning.

1. Run the test with Node 12 (latest supported Node version for CHT 3.x)
2. If it works, write a script to hot patch medic-os to use Node 12. If it doesn't work we'll have to revisit the best way forward here.

**Comment by Hareet** [comment]
@garethbowen Do you think this is related to the intermediary certificate issue we had on older Android versions? We had to chain a cross-signed certificate.

@mrjones-plip The link you posted earlier from [possibly node 8 is affected](https://stackoverflow.com/questions/62107431/curl-error-60-ssl-certificate-problem-certificate-has-expired/62108103#62108103) has a reply that links to [Sectigo External CA Root expiry](https://sectigo.com/knowledge-base/detail/Sectigo-AddTrust-External-CA-Root-Expiring-May-30-2020/kA03l00000117LT). 

If you have time, can you add the cross-signed cert to your let's encrypt test setup? I'll get back to testing this next week

**Comment by mrjones-plip** [comment]
Doing some prep for next week when I might dive into this: checking the `medic-os` image, I see that node is dynamically linked:

```
root@373ae1ffb24b:/srv# file /srv/software/medic-core/v2.1.1/x64/bin/node
/srv/software/medic-core/v2.1.1/x64/bin/node: ELF 64-bit LSB executable, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, for GNU/Linux 2.6.32, BuildID[sha1]=32d31ae138393b3e59b948d01396f80c73b72f0b, stripped
```

If we don't want to release a new version of Medic OS, then I don't think we can use another dynamically linked binary because it will be dependent on libraries that will be ephemerally installed into the container and be lost upon restart.  As such, I believe we need to distribute a statically compiled node 12 (`12.22.12` to be exact as of this writing). 

So maybe the approach is (all untested psuedo code):
1.  [compile node 12 statically](https://github.com/robvanoostenrijk/nodejs-minimal) 
2. put the node binary in a known good place online, this repo is as good as any
3. figure an easy way to find the name of the medic os container - this comes to mind `docker ps --all --filter "name=medic-os" --format '{{.Names}}'`
4. curl it in: `docker exec -it medic-os-1 bash -c "curl https://github.com/medic/cht-core/path-to-node/node12 -o /usr/bin/node /srv/software/medic-core/v2.1.1/x64/bin/node"`
5. make it executable: `docker exec -it medic-os-1 bash -c "chmod +x /srv/software/medic-core/v2.1.1/x64/bin/node`"
6. reboot it: `docker restart medic-os-1`

**Comment by mrjones-plip** [comment]
Steps to test:

1. create CHT 3.x instance with [docker helper](https://docs.communityhealthtoolkit.org/apps/guides/hosting/3.x/app-developer/#cht-docker-helper) 
2. find the name of your `medic-os` container: `docker ps --format '{{.Names}}'`
3. upgrade node: `docker exec -it medic-os-1 bash -c "curl -sSL https://raw.githubusercontent.com/medic/cht-core/7957-node-upgrade-script/scripts/cht-3x-node-upgrade/upgrade.sh | bash"`
4. set up outbound push to a Let's Encrypt host and see if sentinel gives errors: `tail -f /srv/storage/medic-sentinel/logs/medic-sentinel.log`

**Comment by mrjones-plip** [comment]
Remediation plan: since we're not doing a 3.18 release and doing a hot patch instead, unless I hear back from folks, I'll press ahead with:

* add to known issue for applicable versions 3.9 -> 3.17.1
* Add docs page on how to deploy - similar to the [3.9 medic os upgrade](https://github.com/medic/cht-infrastructure/blob/main/docs/3.9_CHT_Upgrade.md), but in docs, instead in GH
* do a post in #product on forums about the hot patch

**Comment by henokgetachew** [comment]
Makes sense. That looks good to me.

**Comment by mrjones-plip** [comment]
@tatilepizs  - feel like doing some bash AT?

**Comment by tatilepizs** [comment]
Thank you @mrjones-plip  for all the detailed explanations and steps to reproduce the issue, it was really easy to follow.

I used [these](https://github.com/medic/cht-core/issues/7957#issuecomment-1350442823) exact steps to reproduce the error, the only difference is that the command to see the sentinel log was missing the `exec -it` section.

```
docker  exec -it medic-os-1 bash -c "tail -f -n100 /srv/storage/medic-sentinel/logs/*.log"
```

Error shown when I try to create a contact:

![image](https://user-images.githubusercontent.com/94494491/212739421-0b6eb5f4-2f67-4a00-b2f2-0cb132a5f279.png)

After upgrading node I saw a success message:

![image](https://user-images.githubusercontent.com/94494491/212741137-4a77e728-41ee-4f9c-9815-7f3b671fe8e4.png)


It is working as expected 🎉

**Comment by mrjones-plip** [comment]
Chat [from slack](https://medic.slack.com/archives/C02F2TAME90/p1674066500549559) we agreed there's no further action needed:

* We do not add further documentation.  This has been around for many MANY months, possibly years, and we're just now encountering it, and only once at that
* We have published a fix for it (this ticket and the [resulting script](https://github.com/medic/cht-core/blob/master/scripts/cht-3x-node-upgrade/upgrade.sh))
* We made hella `Affects:` tags on the ticket so it shows up on the link in release notes for "known issues"
* There is a [forum post](https://forum.communityhealthtoolkit.org/t/post-request-with-outbound/2197?u=mrjones) on the fix
* We should close the ticket and move on \o/

Thanks for the assist @jonathanvq !

**Comment by magp18** [comment]
> #### script sample
> 
> Output of upgrade script looks like this:
> 
> ```
> docker exec -it medic-os-1 bash -c "curl -sSL https://raw.githubusercontent.com/medic/cht-core/7957-node-upgrade-script/scripts/cht-3x-node-upgrade/upgrade.sh | bash"

> ```

The URL for the upgrade sh leads to 404, I think you meant https://raw.githubusercontent.com/medic/cht-core/master/scripts/cht-3x-node-upgrade/upgrade.sh right?

**Comment by mrjones-plip** [comment]
oh no!  Thanks for the heads up @magp18 ! I'll re-open and throw together a quick follow up PR>

**Comment by latin-panda** [comment]
@mrjones-plip should I include this ticket in milestone 4.2.0? For the release notes

**Comment by mrjones-plip** [comment]
hmmmm - I guess so?  There's actually no code for it in the release - just a script that hot patches a live instance of 3.x.  I'll ad it.

thanks again for checking all this!

**Comment by mrjones-plip** [comment]
I should have closed this after submitting the final fix in #8049 - closing now!
